File: lessons/go/stage-1/level-4/solution.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

func main() {
    reader := bufio.NewReader(os.Stdin)
    
    fmt.Println("=== Interactive Greeting Program ===")
    // Get input from user
    fmt.Print("What's your name? ")
    name, _ := reader.ReadString('\n')
    name = strings.TrimSpace(name)  // Remove newline character
    fmt.Println("Hello,", name, "! Nice to meet you!")
    
    fmt.Println()
    fmt.Println("=== Simple Calculator ===")
    // Get two numbers from user
    fmt.Print("Enter first number: ")
    input1, _ := reader.ReadString('\n')
    input1 = strings.TrimSpace(input1)
    num1, err1 := strconv.ParseFloat(input1, 64)
    if err1 != nil {
        fmt.Println("Invalid number entered!")
        return
    }
    
    fmt.Print("Enter second number: ")
    input2, _ := reader.ReadString('\n')
    input2 = strings.TrimSpace(input2)
    num2, err2 := strconv.ParseFloat(input2, 64)
    if err2 != nil {
        fmt.Println("Invalid number entered!")
        return
    }
    
    // Perform calculations
    sum := num1 + num2
    difference := num1 - num2
    product := num1 * num2
    quotient := num1 / num2
    
    // Display results
    fmt.Printf("%.2f + %.2f = %.2f\n", num1, num2, sum)
    fmt.Printf("%.2f - %.2f = %.2f\n", num1, num2, difference)
    fmt.Printf("%.2f * %.2f = %.2f\n", num1, num2, product)
    fmt.Printf("%.2f / %.2f = %.2f\n", num1, num2, quotient)
    
    fmt.Println()
    fmt.Println("=== Age Calculator ===")
    // Ask for birth year and calculate age
    fmt.Print("What year were you born? ")
    birthYearStr, _ := reader.ReadString('\n')
    birthYearStr = strings.TrimSpace(birthYearStr)
    birthYear, err := strconv.Atoi(birthYearStr)
    if err != nil {
        fmt.Println("Please enter a valid year!")
        return
    }
    
    // Simple age calculation (assuming current year is 2025)
    currentYear := 2025
    age := currentYear - birthYear
    fmt.Printf("If you were born in %d, you are about %d years old.\n", birthYear, age)
    
    fmt.Println()
    fmt.Println("=== Simple Quiz ===")
    // Create a simple quiz
    fmt.Print("What is the capital of France? ")
    answer, _ := reader.ReadString('\n')
    answer = strings.TrimSpace(strings.ToLower(answer))
    
    if answer == "paris" {
        fmt.Println("Correct! Well done!")
    } else {
        fmt.Println("Not quite! The answer is Paris.")
    }
    
    fmt.Println()
    fmt.Println("=== Yes/No Question ===")
    // Get a yes/no response (as text)
    fmt.Print("Do you like Go programming? (yes/no): ")
    response, _ := reader.ReadString('\n')
    response = strings.TrimSpace(strings.ToLower(response))
    
    if response == "yes" || response == "y" {
        fmt.Println("Great! Go is a fantastic language!")
    } else {
        fmt.Println("That's okay, everyone has different preferences!")
    }
    
    fmt.Println()
    fmt.Println("=== Multiple Inputs ===")
    // Collect multiple pieces of information
    fmt.Print("What's your favorite color? ")
    color, _ := reader.ReadString('\n')
    color = strings.TrimSpace(color)
    
    fmt.Print("How many pets do you have? ")
    petStr, _ := reader.ReadString('\n')
    petStr = strings.TrimSpace(petStr)
    petCount, err := strconv.Atoi(petStr)
    if err != nil {
        fmt.Println("Please enter a valid number for pets!")
        petCount = 0  // Default value
    }
    
    fmt.Print("What's your favorite food? ")
    food, _ := reader.ReadString('\n')
    food = strings.TrimSpace(food)
    
    fmt.Println()
    fmt.Println("=== Your Profile ===")
    fmt.Printf("Favorite color: %s\n", color)
    fmt.Printf("Number of pets: %d\n", petCount)
    fmt.Printf("Favorite food: %s\n", food)
    fmt.Printf("Thanks for sharing, %s!\n", name)
    
    fmt.Println()
    fmt.Println("=== Input Validation Example ===")
    // Demonstrate input validation
    fmt.Print("Enter a positive number: ")
    userInput, _ := reader.ReadString('\n')
    userInput = strings.TrimSpace(userInput)
    number, err := strconv.ParseFloat(userInput, 64)
    
    if err != nil {
        fmt.Println("Invalid input! That wasn't a number.")
    } else if number > 0 {
        fmt.Printf("Valid positive number: %.2f\n", number)
        fmt.Printf("Its square is: %.2f\n", number*number)
    } else {
        fmt.Println("That's not a positive number!")
    }
}

File: lessons/lua/stage-1/level-4/solution.lua
print("=== Interactive Greeting Program ===")
-- Get input from user
io.write("What's your name? ")  -- io.write() doesn't add a newline
local name = io.read()  -- Read a line from standard input
print("Hello, " .. name .. "! Nice to meet you!")

print()
print("=== Simple Calculator ===")
-- Get two numbers from user
io.write("Enter first number: ")
local num1 = tonumber(io.read())  -- Convert input string to number

io.write("Enter second number: ")
local num2 = tonumber(io.read())  -- Convert input string to number

-- Perform calculations
local sum = num1 + num2
local difference = num1 - num2
local product = num1 * num2
local quotient = num1 / num2

-- Display results
print(num1 .. " + " .. num2 .. " = " .. sum)
print(num1 .. " - " .. num2 .. " = " .. difference)
print(num1 .. " * " .. num2 .. " = " .. product)
print(num1 .. " / " .. num2 .. " = " .. quotient)

print()
print("=== Age Calculator ===")
-- Ask for birth year and calculate age
io.write("What year were you born? ")
local birth_year = tonumber(io.read())
local current_year = tonumber(os.date("%Y"))  -- Get current year
local age = current_year - birth_year

print("If you were born in " .. birth_year .. ", you are about " .. age .. " years old.")

print()
print("=== Simple Quiz ===")
-- Create a simple quiz
io.write("What is the capital of France? ")
local answer = io.read()
if string.lower(answer) == "paris" then
    print("Correct! Well done!")
else
    print("Not quite! The answer is Paris.")
end

print()
print("=== Number Guessing Game ===")
-- A simple number guessing game
math.randomseed(os.time())  -- Seed the random generator
local secret_number = math.random(1, 10)
io.write("I'm thinking of a number between 1 and 10. Can you guess it? ")
local user_guess = tonumber(io.read())

if user_guess == secret_number then
    print("You got it! The number was " .. secret_number)
else
    print("Sorry, the number was " .. secret_number .. ". Better luck next time!")
end

print()
print("=== Multiple Inputs ===")
-- Collect multiple pieces of information
io.write("What's your favorite color? ")
local color = io.read()

io.write("How many pets do you have? ")
local pet_count = tonumber(io.read())

io.write("What's your favorite food? ")
local food = io.read()

print()
print("=== Your Profile ===")
print("Favorite color: " .. color)
print("Number of pets: " .. pet_count)
print("Favorite food: " .. food)
print("Thanks for sharing, " .. name .. "!")

print()
print("=== Input Validation Example ===")
-- Demonstrate input validation
io.write("Enter a positive number: ")
local user_input = io.read()
local number = tonumber(user_input)

if number ~= nil and number > 0 then
    print("Valid positive number: " .. number)
    print("Its square is: " .. number^2)
else
    print("Invalid input! That wasn't a positive number.")
end